### ISSUE TITLE
Timeslider performance is not satisfactory - underlying issues interpolation

### ISSUE THREAD

**Issue by arthurcamara1** [Gapminder/vizabi#164]
Timeslider performance on bubble chart is not satisfactory. If we check the FPS (simple naive check with https://github.com/mrdoob/stats.js/), we see that FPS rate goes as low as 20fps when dragging the slider quickly.

Upon inspection, we can see that  `interpolateValue` is taking the biggest chunk of time. Performance of interpolation should be improved. https://github.com/Gapminder/vizabi/blob/develop/src/base/model.js#L957

*First guess:* we should not interpolate at hook level (each hook will interpolate it's own indicator. 5 hooks = 5 interpolations with 5 bisects). Instead, we should find a way to interpolate all indicators at once. It would reduce the number of interpolations to 1 per frame, regardless of the number of hooks.

**Comment by ghaiklor** [comment]
@arthurcamara1 All what we can do here, it's apply Memoization for interpolateValue function. Quick solution for it - it's just caching of results before. For more optimization we need to re-think a lot of functions here.

**Comment by arthurcamara1** [comment]
Sounds good. So let's re-think those functions :) Which problems can you see?

**Comment by arthurcamara1** [comment]
@ghaiklor I quickly sketched some ideas to make my suggestion above clear.

Suppose we have 4 hooks (4 models hooked to data), like this:
![hookinterpolation](https://cloud.githubusercontent.com/assets/3584238/8374171/75890204-1bf3-11e5-8e95-20546e9b8a0b.jpg)

When we drag the timeslider, the event **change:time** is triggered and we need to update the views, which calls for interpolation in **EACH** of the hooks.

![hookinterpolation 1](https://cloud.githubusercontent.com/assets/3584238/8374184/a1fc947c-1bf3-11e5-96f9-eb5e6230490b.jpg)

Now, each interpolation is expensive, because it **looks for data points before and after the point we want in a potentially big array**. This is basically what it does:

1. *looks for data points before and after the value we want to get.*
2. *linearly interpolates those two points, based on the position of the point we want (fraction of the difference)*

![hookinterpolation 2](https://cloud.githubusercontent.com/assets/3584238/8374281/6e2c6518-1bf4-11e5-8943-90e5143afbe5.jpg)

Now, this is expensive because the bigger the number of hooks, the bigger the number of searches we'll have to perform.

**POSSIBLE SOLUTION**:

the data points between hooks are virtually **the same**. After the first interpolation, we already know where the data point should be located in all hooks (indexBefore, indexNext). We even know the fraction of the difference that we need to add to the the value of the point before to get the interpolated one.

If *those indexes* (and even fractions) could be cached/memoized, all could be reduced to this:

![hookinterpolation 3](https://cloud.githubusercontent.com/assets/3584238/8374380/36948940-1bf5-11e5-8953-be8a47a6360d.jpg)

Then the sequence would be:

- Hook1: Complex Interpolation, looking for right indexes
- Hook2: Simple multiplication using cached indexes 
- Hook3: Simple multiplication using cached indexes 
- Hook4: Simple multiplication using cached indexes 

Does it make sense? @ghaiklor @angieskazka

**Comment by arthurcamara1** [comment]
**Great news.**

I was able to drastically reduce the time of interpolation. I noticed that performance is really improved if we _reduce the call stack_.

When dragging the timeslider, the operations which used to take 60ms now are taking a bit more than 20ms:

![screen shot 2015-07-01 at 12 44 11 am](https://cloud.githubusercontent.com/assets/3584238/8443877/e7163438-1f8a-11e5-9d5c-c42a3dbd98f6.png)

Explained Reason:

This:
```
for(i=0; i<manyLists.length; i++) {
    result.push(interpolate(manyLists[i].arr, 757.87));
}
```

is much faster (2.5X) than this:
```
for(i=0; i<manyLists.length; i++) {
    result.push(manyLists[i].get(757.87));
}
```

I wrote an example here: http://jsfiddle.net/arthurcamara/8wq3rcfu/

Of course, in core the fix is much more complex than that, but that's the general idea. **Check out the code changes here:** c7c3f762daa99769fa5a39e67ba4d84f28704909

We can still improve things, but we're getting very very close to have much better performance. I'll consider this closed.
